### PR TITLE
fix: upgrade JAX from 0.8.1 to 0.9.2

### LINKED ISSUE #1025: [Feature] Upgrade JAX dependency from 0.8.1 to the newest version
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve it, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

## Summary

Upgrade JAX from 0.8.1 to 0.9.2, and libtpu from 0.0.30 to 0.0.38.

Relates to #909 — "Upgrade to latest JAX version"

Reference: [tpu-inference](https://github.com/vllm-project/tpu-inference) already uses JAX 0.9.2 + libtpu 0.0.39.

## Current State

| Package | Current | Target |
|---------|---------|--------|
| jax     | 0.8.1   | 0.9.2  |
| jaxlib  | 0.8.1   | 0.9.2  |
| libtpu  | 0.0.30  | 0.0.38 |

Version pins in `python/pyproject.toml` (tpu/gpu/cpu/all extras).

Known blocker: #584 (logprob regression on JAX 0.8.2)

## Plan

- [ ] Step 1: Update version pins in `python/pyproject.toml`
- [ ] Step 2: Fix CI failures from API incompatibilities
- [ ] Step 3: Verify model accuracy (MiMo-V2-Flash on AIME test set)

## Known Breaking Changes (0.8.1 → 0.9.2)

| Version | Change | Impact |
|---------|--------|--------|
| 0.8.2 | Tracer no longer inherits from jax.Array | Affects #584 (logprob path) |
| 0.8.2 | `jax.interpreters.pxla` fully deprecated | Not used, no impact |
| 0.8.2 | `jax.lax.pvary` deprecated | Not used, no impact |
| 0.9.1 | shard_map enforces PartitionSpec matching in Explicit mode | Affects 15+ Pallas kernel call sites |

## Local Import Test

All sglang-jax modules import successfully on JAX 0.9.2 (local, no TPU). Issues expected at runtime (JIT compilation, shard_map enforcement) — requires TPU VM for validation.

## Work Estimate

~10-15 days. Major risks:
- #584 root cause not yet identified
- shard_map validation across all Pallas kernels
- Potential accuracy regression from XLA optimization changes

### PR DESCRIPTION
**Pull request by Aharrypotter**
## Summary

- Upgrade JAX/JAXlib from 0.8.1 to 0.9.2 and libtpu from 0.0.30 to 0.0.38.
- Fix JAX 0.9.1+ strict `shard_map` `in_specs` enforcement across FlashAttention, MLA, KV cache, linear attention, and fused MoE paths.
- Fix JAX 0.9.2 compatibility issues around `reshape` `out_sharding`, `pallas_call` Manual mesh axes, and the `jax.reshard` API namespace.
- Add sharding-aware helper paths that derive shard specs from actual `NamedSharding` / tracer sharding instead of relying on divisibility heuristics.
- Fix FlashAttention GQA and `attention_sink` sharding behavior under local head-ratio constraints.
- Harden LoRA batching with explicit batch metadata, stable base-model sentinel handling, no-LoRA row preservation, clearer slot accounting, and stronger runtime validation.
- Remove confirmed dead LoRA helper code and tighten backend contracts.
- Stabilize LoRA debug dump request ordering and padded-batch `lora_ids` handling for DP-aware batches.

Closes #1025

## Motivation

JAX 0.9.1+ tightened several sharding contracts that were previously permissive. In particular, `shard_map` now requires `in_specs` to match the actual input sharding instead of silently resharding mismatched inputs. JAX 0.9.2 also exposes stricter behavior around `reshape(..., out_sharding=...)`, Pallas mesh axes, and reshard API placement.

This PR brings SGLang-JAX in line with the newer JAX/libtpu stack used by TPU inference work, while preserving model accuracy and making LoRA batch handling explicit enough to catch invalid runtime state early.

### Core JAX 0.9.2 sharding compatibility

**FlashAttention / MLA / linear attention**
- Add effective-axis helper logic that reads the actual input sharding, including tracer `aval.sharding` inside `jax.jit`.
- Derive `shard_map` `in_specs` / `out_specs` from actual tensor, DP, cache, sequence-length, and metadata sharding.
- Fix FlashAttention GQA behavior when local Q heads are not divisible by local KV heads by resharding Q to replicated where required.
- Fix `attention_sink` sharding by deriving from the actual Q sharding and resharding to replicated when Q is replicated on the tensor axis.

**KV cache / Pallas**
- Add local/test-safe update paths and wrap only multi-device paths with `shard_map`.
- Strip explicit `NamedSharding` where Pallas kernels require Manual mesh axes.
- Convert host-side metadata with `jax.device_get` + numpy where required.

**Fused MoE**
- Replace the old `jax.sharding.reshard` namespace with `jax.reshard`.
- Update tests to avoid relying on removed or unstable JAX internal APIs.

### LoRA correctness and validation

**Batch metadata model**
- Introduce `LoRABatchPlan` as explicit CPU-side LoRA planning metadata:
  - request-granular `weight_indices`
  - slot-granular `ranks_by_slot`
  - slot-granular `scalings_by_slot`
- Validate invalid slot indices, negative ranks, negative scaling, and non-finite scaling before backend expansion.
- Expand request-level metadata to token-level metadata in the backend with explicit EXTEND and DECODE checks.

**Base-model and no-LoRA handling**
- Reserve and normalize the base-model LoRA sentinel consistently.
- Preserve base-model rows exactly in mixed LoRA batches when the rank is zero.
- Reject unloaded dynamic adapters instead of silently falling back to base-model output.
- Reject `None` adapter loads into non-base LoRA slots.

**Backend and memory-pool hardening**
- Align `BaseLoRABackend` method signatures with the BGMV implementation and layer call sites.
- Raise `NotImplementedError` for unimplemented backend methods.
- Validate EXTEND metadata length, negative `extend_seq_lens`, and token-padding invariants.
- Add contextual errors for missing memory-pool slots and invalid buffer zeroing inputs.
- Avoid mutating scheduler-owned `model_worker_batch.lora_ids`; normalization is now local to LoRA planning.
- Remove confirmed dead LoRA helper code and normalize `jnp.concatenate` usage.

### Scheduler / infrastructure

- Add `arrival_index` / `output_order_selector` handling to keep debug dumps stable.
- Write `lora_ids` per padded batch slot in DP-aware scheduling paths.
- Normalize cached `/models/` paths in the model loader.
- Add subset device-mesh support by truncating devices to the requested mesh shape.

## Accuracy Tests

LoRA accuracy was checked against Hugging Face reference outputs. Dense and MoE MMLU checks were also run to cover non-LoRA serving paths affected by the JAX 0.9.2 sharding changes.

| Area | Command | What it validates | Result |
| --- | --- | --- | --- |
| LoRA HF alignment | `python test/srt/lora/test_align_lora_accuracy.py` | LoRA prefill/decode logits vs Hugging Face reference | PASS, `Ran 3 tests in 282.124s` |
| Dense MMLU | `python test/srt/run_suite.py --suite accuracy-test-tpu-v6e-1` | Non-LoRA Qwen/Qwen3-8B serving with FlashAttention | PASS, score `0.735`, total latency `185.960s` |
| MoE MMLU | `python -m unittest test_qwen3_moe_models.TestQwenModel.test_mmlu` | Qwen/Qwen3-30B-A3B serving with `moe_backend=epmoe`, `tp_size=4`, `ep_size=4` | PASS, score `0.578`, total latency `51.503s` |

LoRA alignment details:

| Scenario | Prefill max diff | Prefill mean diff | Decode max diff | Decode mean diff |
| --- | ---: | ---: | ---: | ---: |
| Single prompt | `2.193451e-05` | `3.758544e-06` | `1.907349e-05` | `2.803902e-06` |
| Multi prompt | N/A | N/A | `4.255009e-02` / `9.037256e-02` / `5.063438e-02` | `7.126036e-03` / `2.825490e-02` / `1.182749e-02` |

The multi-prompt values above are absolute logit diffs, not percentages. The test uses `np.testing.assert_allclose(..., atol=9e-2, rtol=9e-2)`, so the largest observed absolute diff (`9.037256e-02`) passes the combined allclose tolerance. The mean diffs stay in the `7e-3` to `2.8e-2` range.

The trailing `multiprocessing.resource_tracker KeyError` after the `OK` line is cleanup noise from the test process; the command exits with status 0.

### Validation matrix

| Category | Command | Hardware / mode | Status | Notes |
| --- | --- | --- | --- | --- |
| Syntax/import sanity | `python -m compileall -q python/sgl_jax/srt/lora test/srt/lora/test_bgmv_backend.py test/srt/lora/test_lora_manager_optimization.py` | CPU host | PASS | LoRA modules and tests compile cleanly |
| LoRA CPU logic | `JAX_PLATFORMS=cpu python -m unittest test.srt.lora.test_lora_manager_optimization test.srt.lora.test_bgmv_backend.TestRankZeroOutputRestore` | CPU | PASS | `14 tests OK` |
| LoRA TPU backend | `python test/srt/lora/test_bgmv_backend.py` | TPU | PASS | `14 tests OK` |
| LoRA accuracy | `python test/srt/lora/test_align_lora_accuracy.py` | TPU | PASS | `3 tests OK`; HF logit alignment checked |
| Dense non-LoRA accuracy | `python test/srt/run_suite.py --suite accuracy-test-tpu-v6e-1` | TPU v6e-4 environment | PASS | Qwen/Qwen3-8B MMLU score `0.735` |
| MoE E2E accuracy | `python -m unittest test_qwen3_moe_models.TestQwenModel.test_mmlu` | TPU v6e-4, `moe_backend=epmoe` | PASS | Qwen/Qwen3-30B-A3B MMLU score `0.578` |
| Fused MoE E2E attempt | `python -m unittest test_qwen3_moe_models.TestQwenModelForFusedMoE.test_mmlu` | TPU v6e-4, `moe_backend=fused` | BLOCKED | Fails during fused MoE EXTEND precompile: `Expected intermediate_size=768 to be aligned to bf=512` |

The fused MoE result is a current fused kernel block-config/model compatibility limitation for Qwen3-MoE in this environment. It fails before request-level accuracy execution, is not counted as passing evidence for this PR, and does not block the validated default `epmoe` MoE path.

### Unit test suite sharding

On the current TPU v6e-4 environment:

```bash
source .venv/bin/activate
python test/srt/run_suite.py --suite unit-test-tpu-v6e-1 --auto-partition-id 0 --auto-partition-size 3
python test/srt/run_suite.py --suite unit-test-tpu-v6e-1 --auto-partition-id 1 --auto-partition-size 3
python test/srt/run_suite.py --suite unit-test-tpu-v6e-1 --auto-partition-id 2 --auto-partition-size 3
```

Observed:

| Shard | Status | Notes |
| --- | --- | --- |
| `--auto-partition-id 0 --auto-partition-size 3` | FAILED | Non-LoRA `python/sgl_jax/test/test_flashattention.py::test_gqa_prefill_with_custom_mask` hits TPU compile-time VMEM OOM on v6e-4: `64.15M > 64.00M` |
| `--auto-partition-id 1 --auto-partition-size 3` | PASS | Completed successfully |
| `--auto-partition-id 2 --auto-partition-size 3` | PASS | Completed successfully |

The shard-0 failure is a FlashAttention TPU compiler VMEM limit in the current TPU v6e-4 environment. It is tracked here as an environment/compiler limitation and has not been verified on a larger slice in this validation pass.

## Benchmarking and Profiling

No dedicated throughput/latency benchmark is included in this PR. The primary validation focus is dependency compatibility, sharding correctness, and LoRA correctness hardening. This PR does touch sharding and metadata paths that can affect compilation behavior, so a focused `bench_one_batch.py` comparison would be a useful follow-up if performance sign-off is required.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by Aharrypotter** [comment]
cc @Bob-Chen222 @hlFu @pengchengneo

**Comment by Aharrypotter** [comment]
@claude

File: python/sgl_jax/srt/lora/backend/bgmv_backend.py [hunk 1]
```diff
@@ -365,10 +389,7 @@ def bgmv_expand_slice(
     pad_right = output_array.shape[-1] - (slice_offset + slice_size)
     outputs = jnp.pad(outputs, ((0, 0), (pad_left, pad_right)), mode="constant", constant_values=0)
 
-    if output_array is not None:
-        return output_array + outputs
-    else:
-        return outputs
+    return output_array + outputs
```
Comment (hlFu) [inline python/sgl_jax/srt/lora/backend/bgmv_backend.py:392]: Does this change the original logic?

File: python/sgl_jax/srt/layers/attention/flashattention_backend.py [hunk 2]
```diff
@@ -537,43 +537,112 @@ def __call__(
...
+        v_data = effective_axis(v_arg, 0, dpa)
+        v_axis = effective_axis(v_arg, 1, axis)
+        # GQA head ratio: if q is tensor-sharded, verify local heads are valid.
+        if q_axis is not None:
+            axis_size = self.mesh.shape.get(axis, 1) if self.mesh else 1
+            local_q = q_arg.shape[1] // axis_size
+            local_kv = k_arg.shape[1] // axis_size if kv_axis is not None else k_arg.shape[1]
+            if local_q % local_kv != 0:
+                if self.mesh is not None and q_data is not None:
+                    q_arg = jax.reshard(q_arg, NamedSharding(self.mesh, P(q_data, None, None)))
```
Comment (hlFu) [inline python/sgl_jax/srt/layers/attention/flashattention_backend.py:566]: Is this necessary to avoid shard_map break?

File: python/sgl_jax/srt/layers/attention/linear/gdn_backend.py [hunk 3]
```diff
@@ -200,9 +201,25 @@ def forward_decode(
         """One token per request — single conv1d update + parallel single
         recurrence step across the batch, all inside a shard_map."""
         state_indices = self.forward_metadata.recurrent_indices
-        tp = _mesh_tp_size(self.mesh)
+        mixed_data = effective_axis(mixed_qkv, 0, "data")
```
Comment (hlFu) [inline python/sgl_jax/srt/layers/attention/linear/gdn_backend.py:204]: This seems beyond the scope of upgrading JAX. This is introducing DP. Did you involve this change because of rebasing main branch?
